### PR TITLE
[SD-172] Added support for CAPTCHAs in webforms

### PR DESCRIPTION
**Pull request by jeffdowdle**
**Issue**: https://digital-vic.atlassian.net/browse/SD-172

Screenshot 2024-09-16 at 2 44 06 PM

### What I did

- Added support for CAPTCHAs in webforms, support for:
  - reCAPTCHA v2 (checkbox) and v3
  - Turnstile
- Wrote a confluence doc explaining the solution https://digital-vic.atlassian.net/wiki/spaces/RPL/pages/2844688488/CAPTCHA+support
- Adds support for CAPTCHAs in content rating form
- Supports multiple CAPTCHAs of the same and mixed types on the same page

### How to test

- Follow instructions here: https://digital-vic.atlassian.net/wiki/spaces/RPL/pages/2844688488/CAPTCHA+support
- It's unfortunately very difficult to write automated tests for this

#### For all PR's

- [ ] I've added relevant changes to the project Readme if needed.
- [ ] I've updated the documentation site as needed.
- [ ] I have added unit tests to cover my changes (if not applicable, please state why in a comment)

#### For new components only

- [ ] I have added a story covering all variants
- [ ] I have checked a11y tab in storybook passes
- [ ] Any events are emitted on the event bus

### CODE REVIEW

File: packages/ripple-ui-forms/src/components/RplFormActions/RplFormActions.vue [hunk 1]
```diff
@@ -63,8 +63,13 @@ const handleReset = () => {
     { global: props.globalEvents }
   )
 }
+
+const captchaElementId = computed(() => {
+  return getCaptchaElementId(form?.id)
+})
 </script>
 <template>
+  <div v-if="captchaElementId" :id="captchaElementId"></div>
```
Comment (jeffdowdle) [inline packages/ripple-ui-forms/src/components/RplFormActions/RplFormActions.vue:72]: If applicable, the captcha checkbox/badge will be rendered here

File: packages/ripple-tide-webform/utils/captcha.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    case CaptchaType.RECAPTCHA_V2:
+      return getGoogleRecaptchaV2Response(_window)
+    case CaptchaType.TURNSTILE:
+      return getCloudfareTurnstileResponse(_window)
+    default:
+      throw new Error(`Captcha type not implemented: '${captchaConfig.type}'`)
+  }
+}
+
+export const getCaptchaResponse = async (
```
Comment (jeffdowdle) [inline packages/ripple-tide-webform/utils/captcha.ts:36]: This is called on submit of the form to get the captcha token

File: packages/ripple-tide-webform/composables/useCaptchaWidget.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    case CaptchaType.TURNSTILE:
+      if (_window && _window.turnstile) {
+        _window.turnstile.reset(widgetId)
+      }
+      break
+    default:
+  }
+}
+
+export const useCaptchaWidget = (
```
Comment (jeffdowdle) [inline packages/ripple-tide-webform/composables/useCaptchaWidget.ts:85]: This loads the third party scripts and calls any nessacary initialisation code for the captcha type

File: packages/ripple-tide-webform/composables/use-webform-submit.ts [hunk 4]
```diff
@@ -29,7 +37,8 @@ export function useWebformSubmit(formId: string) {
         site: config.tide.site
       },
       headers: {
-        'Content-Type': 'application/vnd.api+json;charset=UTF-8'
+        'Content-Type': 'application/vnd.api+json;charset=UTF-8',
+        'x-captcha-response': maybeCaptchaResponse || undefined
```
Comment (jeffdowdle) [inline packages/ripple-tide-webform/composables/use-webform-submit.ts:41]: The captcha token is sent via a header to the nuxt server endpoint

File: packages/nuxt-ripple/server/utils/verifyCaptcha.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    case CaptchaType.RECAPTCHA_V2:
+      return await verifyGoogleRecaptchaV2(secretKey, captchaResponse)
+    case CaptchaType.TURNSTILE:
+      return await verifyCloudfareTurnstile(secretKey, captchaResponse)
+    default:
+      return false
+  }
+}
+
+const verifyCaptcha = async (event: H3Event) => {
```
Comment (jeffdowdle) [inline packages/nuxt-ripple/server/utils/verifyCaptcha.ts:120]: This is where we take the token generated in the browser and validate it which the CAPTCHA provider, this needs to happen server side.

We have to get the webform from drupal here to check if the captcha is enabled or not.

**Comment by jeffdowdle** [comment]
Still waiting for backend to be finalised, but this is ready to start looking at @dylankelly

File: packages/nuxt-ripple/components/TideContentRating.vue [hunk 6]
```diff
@@ -17,6 +17,16 @@ const contentRatingFormId = 'tide_webform_content_rating'
 const isMounted = ref(false)
 const pageUrl = ref('')
 
+const { data: webformData, status: webformFetchStatus } = await useFetch(
```
Comment (lambry) [inline packages/nuxt-ripple/components/TideContentRating.vue:20]: Does `useWebformSchema` work here?

**Pull request merged**